### DISCUSSION TITLE
How to call command from a data template?

### DISCUSSION THREAD

**Discussion by Kirk-iOS** [dotnet/maui#17771] [Q&A] [answered]
Hello,

I am building an application using the MVVM pattern. In one of my pages I have a carousel view which I have defined the data template directly in XAML.

One of the views in the data template is a button which navigates to another page. I have created a ICommand in the viewmodel which navigates to this page and I am attempting to run the command when the button is clicked. 

How does the data binding work here? As far as I can tell the data template has no visibility of the parents page BindingContext. If anyone can link me any documentation that covers this or explain where I am going wrong, that would be great.

Thanks.

**Comment by sjordanGSS** [comment]
I use RelativeBinding to do this sort of thing. RelativeBinding with Mode=AncestorBindingContext will traverse the view hierarchy looking for the viewmodel by type as defined in AncestorType, and allow you to bind to properties on that viewmodel. Sadly you don't get intellisense using this method, but it does work.

**Reply by Kirk-iOS** [reply]
Thanks I will give it ago. Do you get intellisense if you set the data type? x:DataType="ViewModel"

**Reply by sjordanGSS** [reply]
possibly? But that would mess up the Intellisense on any other bindings you have on that object. Better to just be careful with your relative bindings.

**Reply by matmork** [reply]
`Command="{Binding MyCommand, Source={RelativeSource AncestorType={x:Type viewModels:MyViewModel}}}"`

**Comment by nor0x** [comment, marked as answer]
you could also use x:Reference in the Binding Source and reference the element which has the ViewModel as a BindingContext

given you have a ContentPage with a Command in the BindingContext

```xml
<ContentPage
    xmlns="http://schemas.microsoft.com/dotnet/2021/maui"
    xmlns:x="http://schemas.microsoft.com/winfx/2009/xaml"
    xmlns:vm="clr-namespace:MyProject.ViewModels"
    x:Name="PageSelf"
    x:DataType="vm:MyViewModel"
    x:Class="MyProject.MyPage">
```

...
```xml
<!-- inside DataTemplate -->
Command="{Binding Source={x:Reference PageSelf}, Path=BindingContext.MyCommand}"
```

**Reply by Kirk-iOS** [reply]
This works nicely.

**Reply by sisaacks** [reply]
does not work in .net 9 if you declare a datatype on the template

**Reply by amadzarak** [reply]
> does not work in .net 9 if you declare a datatype on the template

Hey there, appreciate the tip. I seem to be also having some issues with commanding from a data template. Do you have any pointers on where I can find more information about this?

**Reply by codychaplin** [reply]
I recently figured this out [here](https://stackoverflow.com/questions/79544099/net-9-maui-how-to-access-command-from-vm-from-within-a-datatemplate/79544106#79544106). You have to use a relative binding in .NET 9.
